### PR TITLE
TOOL-1303 Optimzed caching hash calculations.

### PR DESCRIPTION
**Pull request by richard-bogdan-bitrise**
Added measurements: https://bitrise.atlassian.net/browse/TOOL-1303

### CODE REVIEW

File: cache_path.go [hunk 1]
```diff
@@ -201,26 +204,27 @@ func match(pth string, excludeByPattern map[string]bool) (bool, bool) {
...
-		skip, exclude := match(pth, excludeByPattern)
-		if exclude {
-			// this file should not be included in the cache
-			continue
+			if skip || indicator == "-" {
+				// this file's changes does not fluctuates existing cache invalidation
+				indicator = ""
+			} else if len(indicator) == 0 {
+				// the file's own content fluctuates existing cache invalidation
+				indicator = pth
```
Comment (lpusok) [inline cache_path.go]: Mutating the map key has unwanted effects. Consider the test case:
```	{
			name:                "no indicator, own content is the indicator",
			indicatorMap:        map[string][]string{"": {"path/to/cache", "path/to/cache2"}},
			excludeByPattern:    map[string]bool{},
			indicatorByCachePth: map[string][]string{"path/to/cache": {"path/to/cache"}, "path/to/cache2": {"path/to/cache2"}},
			wantErr:             false,
		},
        ```

File: cache_descriptor.go [hunk 2]
```diff
@@ -72,27 +70,36 @@ func compare(old map[string]string, new map[string]string) (r result) {
...
-// cacheDescriptor creates a cache descriptor for a given cache_path - change_indicator_path mapping.
-func cacheDescriptor(indicatorByCachePth map[string]string, method ChangeIndicator) (map[string]string, error) {
-	descriptor := map[string]string{}
-	for pth, indicatorPth := range indicatorByCachePth {
-		if len(indicatorPth) == 0 {
-			// this file's changes does not fluctuates existing cache invalidation
-			descriptor[pth] = "-"
-			continue
+// transformToIndicatorByPth transforms an indicator map to an indicatorByPth map.
+func transformToIndicatorByPth(indicatorMap map[string][]string) map[string]string {
```
Comment (lpusok) [inline cache_descriptor.go]: This is a duplicate of convertDescriptorToIndicatorByPath.

File: cache_path.go [hunk 3]
```diff
@@ -135,32 +135,35 @@ func normalizeIndicatorByPath(indicatorByPath map[string]string) (map[string]str
...
-			normalized[file] = indicator
-		}
-		for _, file := range symlinkPaths {
-			// this file's changes does not fluctuates existing cache invalidation
-			normalized[file] = "-"
+			regularFiles, symlinkPaths, err := expandPath(pth)
+			if err != nil {
+				return nil, err
+			}
+			for _, file := range regularFiles {
```
[the remark this answers is not in the document]
Comment (richard-bogdan-bitrise) [reply cache_path.go]: Updated.

File: cache_descriptor.go (hunk 2 above)
Comment (richard-bogdan-bitrise) [reply cache_descriptor.go]: Removed.

File: cache_path.go (hunk 1 above)
Comment (richard-bogdan-bitrise) [reply cache_path.go]: Updated code.

**Comment by lpusok** [comment]
Paths got mixed up:
Before:
```
+------------------------------------------------------------------------------+
| (4) Push reference cache                                                     |
+------------------------------------------------------------------------------+
| id: ./                                                                       |
| version:                                                                     |
| collection: path                                                             |
| toolkit: go                                                                  |
| time: 2019-11-13T14:58:17+01:00                                              |
+------------------------------------------------------------------------------+
|                                                                              |
Config:
- Paths: ./node_modules -> ./yarn.lock


- IgnoredPaths: 

- CacheAPIURL: file://Users/lpusok/Develop/go/src/github.com/bitrise-steplib/steps-cache-push/cachetest
- FingerprintMethodID: file-content-hash
- CompressArchive: false
- DebugMode: true
- StackID: 

Cleaning paths
Done in 592.962357ms

Checking previous cache status
No previous cache info found
Done in 37.713299086s

Generating cache archive
Done in 3.169922288s

Uploading cache archive
Done in 3.028649088s

Total time: 44.505107643s
|                                                                              |
+---+---------------------------------------------------------------+----------+
| ✓ | Push reference cache                                          | 1.0 min  |
+---+---------------------------------------------------------------+----------+
```
After:
```
+------------------------------------------------------------------------------+
| (4) Push reference cache                                                     |
+------------------------------------------------------------------------------+
| id: ./                                                                       |
| version:                                                                     |
| collection: path                                                             |
| toolkit: go                                                                  |
| time: 2019-11-13T15:04:57+01:00                                              |
+------------------------------------------------------------------------------+
|                                                                              |
Config:
- Paths: ./node_modules -> ./yarn.lock


- IgnoredPaths: 

- CacheAPIURL: file://Users/lpusok/Develop/go/src/github.com/bitrise-steplib/steps-cache-push/cachetest
- FingerprintMethodID: file-content-hash
- CompressArchive: false
- DebugMode: true
- StackID: 

Cleaning paths
indicator is a directory: /Users/lpusok/Develop/go/src/github.com/bitrise-steplib/steps-cache-push/_tmp1/node_modules
Done in 29.355µs

No path to cache, skip caching...
|                                                                              |
+---+---------------------------------------------------------------+----------+
| ✓ | Push reference cache                                          | 17.94 sec|
+---+---------------------------------------------------------------+----------+
```

File: cache_archive.go [hunk 4]
```diff
@@ -53,8 +53,8 @@ func NewArchive(pth string, compress bool) (*Archive, error) {
 }
 
 // Write writes the given files in the cache archive.
-func (a *Archive) Write(pths []string) error {
-	for _, pth := range pths {
+func (a *Archive) Write(pathToIndicator map[string]string) error {
```
Comment (godrei) [inline cache_archive.go]: why did we change this function's interface? it is still using only the path

File: cache_descriptor.go [hunk 5]
```diff
@@ -39,62 +39,69 @@ func (r result) hasChanges() bool {
 }
 
 // compare compares two cache descriptor file and return the differences.
-func compare(old map[string]string, new map[string]string) (r result) {
+func compare(old map[string]string, new map[string]string) result {
```
Comment (godrei) [inline cache_descriptor.go]: why did we do this change?

File: cache_descriptor.go [hunk 6]
```diff
@@ -39,62 +39,69 @@ func (r result) hasChanges() bool {
...
+		indicatorToPaths[indicatorPath] = append(indicatorToPaths[indicatorPath], path)
+	}
+
+	for indicatorPath, paths := range indicatorToPaths {
 		var indicator string
-		if method == MD5 {
-			indicator, err = fileContentHash(indicatorPth)
+		var err error
+		if len(indicatorPath) == 0 {
+			// this file's changes does not fluctuates existing cache invalidation
```
Comment (godrei) [inline cache_descriptor.go]: `fluctuates` -> fluctuate

File: cache_descriptor.go (hunk 5 above)
Comment (lpusok) [reply cache_descriptor.go]: it is not recommended (linters also call out) long functions that naked return values.

File: cache_archive.go (hunk 4 above)
Comment (lpusok) [reply cache_archive.go]: Seemed unnecessary to copy over the paths, can not see performance difference tough.

**Pull request merged**